File: src/main/java/dev/thatredox/chunkynative/common/export/ResourcePalette.java
package dev.thatredox.chunkynative.common.export;

public interface ResourcePalette<T> {
    /**
     * Add a resource to the palette and get the reference.
     */
    int put(T resource);

    /**
     * Release this resource palette. The default implementation does nothing, but palettes with native
     * resources can use this to free those resources.
     */
    default void release() {}
}

File: src/main/native/src/resource_palette.rs
pub struct ResourcePalette {
    pub resources: Vec<Vec<u32>>,
}

impl ResourcePalette {
    pub fn new() -> Self {
        ResourcePalette { resources: vec![] }
    }

    pub fn put(&mut self, resource: Vec<u32>) -> usize {
        let index = self.resources.len();
        self.resources.push(resource);
        index
    }
}
